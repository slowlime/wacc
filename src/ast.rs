use crate::token::{Symbol, Token};

macro_rules! define_op_kind {
    ($name:ident { $( $op:ident => $symbol:ident, )+ }) => {
        #[derive(Debug, Clone, Copy, Eq, PartialEq, Hash)]
        pub enum $name {
            $( $op, )+
        }

        impl $name {
            pub fn symbol(self) -> Symbol {
                match self {
                    $( Self::$op => Symbol::$symbol, )+
                }
            }
        }

        impl TryFrom<Symbol> for $name {
            type Error = ();

            fn try_from(symbol: Symbol) -> Result<Self, Self::Error> {
                match symbol {
                    $( Symbol::$symbol => Ok(Self::$op), )+
                    _ => Err(()),
                }
            }
        }
    }
}

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub struct Program<'a> {
    pub classes: Vec<Class<'a>>,
}

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub struct Class<'a> {
    pub name: Name<'a>,
    pub inherits: Option<Name<'a>>,
    pub features: Vec<Feature<'a>>,
}

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub struct Name<'a>(pub Token<'a>);

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub enum Feature<'a> {
    Method(Method<'a>),
    Field(Binding<'a>),
}

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub struct Method<'a> {
    pub name: Name<'a>,
    pub params: Vec<Formal<'a>>,
    pub return_ty: Name<'a>,
    pub body: Expr<'a>,
}

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub struct Formal<'a> {
    pub name: Name<'a>,
    pub ty: Name<'a>,
}

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub struct Binding<'a> {
    pub name: Name<'a>,
    pub ty: Name<'a>,
    pub init: Option<Expr<'a>>,
}

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub enum Expr<'a> {
    Assignment(Assignment<'a>),
    Call(Call<'a>),
    If(If<'a>),
    While(While<'a>),
    Block(Block<'a>),
    Let(Let<'a>),
    Case(Case<'a>),
    New(New<'a>),
    BinOp(BinOpExpr<'a>),
    UnOp(UnOpExpr<'a>),
    Name(Name<'a>),
    Int(IntLit<'a>),
    String(StringLit<'a>),
    Bool(BoolLit<'a>),
}

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub struct Assignment<'a> {
    pub name: Name<'a>,
    pub expr: Box<Expr<'a>>,
}

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub struct Call<'a> {
    pub receiver: Receiver<'a>,
    pub method: Name<'a>,
    pub args: Vec<Expr<'a>>,
}

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub enum Receiver<'a> {
    SelfType,
    Dynamic(Box<Expr<'a>>),
    Static {
        object: Box<Expr<'a>>,
        ty: Name<'a>,
    },
}

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub struct If<'a> {
    pub condition: Box<Expr<'a>>,
    pub consequent: Box<Expr<'a>>,
    pub alternative: Box<Expr<'a>>,
}

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub struct While<'a> {
    pub condition: Box<Expr<'a>>,
    pub body: Box<Expr<'a>>,
}

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub struct Block<'a> {
    pub body: Vec<Expr<'a>>,
}

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub struct Let<'a> {
    pub bindings: Vec<Binding<'a>>,
    pub expr: Box<Expr<'a>>,
}

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub struct Case<'a> {
    pub scrutinee: Box<Expr<'a>>,
    pub arms: Vec<CaseArm<'a>>,
}

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub struct CaseArm<'a> {
    pub name: Name<'a>,
    pub ty: Name<'a>,
    pub expr: Box<Expr<'a>>,
}

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub struct New<'a>(pub Name<'a>);

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub struct BinOpExpr<'a> {
    pub op: BinOpKind,
    pub lhs: Box<Expr<'a>>,
    pub rhs: Box<Expr<'a>>,
}

define_op_kind!(BinOpKind {
    Add => Plus,
    Subtract => Minus,
    Multiply => Asterisk,
    Divide => Slash,
    LessThan => Less,
    LessEquals => LessEquals,
    Equals => Equals,
});

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub struct UnOpExpr<'a> {
    pub op: UnOpKind,
    pub expr: Box<Expr<'a>>,
}

define_op_kind!(UnOpKind {
    IsVoid => IsVoid,
    Complement => Tilde,
    Not => Not,
});

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub struct IntLit<'a>(pub Token<'a>);

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub struct StringLit<'a>(pub Token<'a>);

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub struct BoolLit<'a>(pub Token<'a>);
