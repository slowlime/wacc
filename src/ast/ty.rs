use std::borrow::Cow;
use std::fmt::{self, Display};

use serde::Serialize;

use crate::try_match;
use crate::util::{slice_formatter, CloneStatic};

use super::TyName;

fn serialize_bytes_as_string<S>(bytes: &[u8], ser: S) -> Result<S::Ok, S::Error>
where
    S: serde::Serializer,
{
    String::from_utf8_lossy(bytes).serialize(ser)
}

/// A potentially-unresolved type.
#[derive(Serialize, Debug, Clone, PartialEq, Eq, Hash)]
pub enum Ty<'buf> {
    Unresolved(UnresolvedTy<'buf>),
    Resolved(ResolvedTy<'buf>),
}

impl CloneStatic<Ty<'static>> for Ty<'_> {
    fn clone_static(&self) -> Ty<'static> {
        match self {
            Self::Unresolved(ty) => Ty::Unresolved(ty.clone_static()),
            Self::Resolved(ty) => Ty::Resolved(ty.clone_static()),
        }
    }
}

impl<'buf> From<UnresolvedTy<'buf>> for Ty<'buf> {
    fn from(ty: UnresolvedTy<'buf>) -> Self {
        Self::Unresolved(ty)
    }
}

impl<'buf> From<ResolvedTy<'buf>> for Ty<'buf> {
    fn from(ty: ResolvedTy<'buf>) -> Self {
        Self::Resolved(ty)
    }
}

#[derive(Serialize, Debug, Clone, PartialEq, Eq, Hash)]
pub enum UnresolvedTy<'buf> {
    Named(TyName<'buf>),

    Function {
        args: Vec<Ty<'buf>>,
        ret: Box<Ty<'buf>>,
    },

    SelfType,
}

impl CloneStatic<UnresolvedTy<'static>> for UnresolvedTy<'_> {
    fn clone_static(&self) -> UnresolvedTy<'static> {
        match self {
            Self::Named(name) => UnresolvedTy::Named(name.clone_static()),
            Self::Function { args, ret } => UnresolvedTy::Function {
                args: args.clone_static(),
                ret: Box::new(ret.clone_static()),
            },
            Self::SelfType => UnresolvedTy::SelfType,
        }
    }
}

pub trait TyExt<'buf>: private::Sealed {
    /// Returns:
    /// - `None` if the type is resolved
    /// - `Some(None)` if the type is unresolved and has to be inferred
    /// - `Some(Some(name))` if the type is unresolved and explicitly named
    fn raw(&self) -> Option<Option<&UnresolvedTy<'buf>>>;

    /// If the type is resolved as `t`, returns `Some(t)`.
    /// Returns `None` otherwise.
    fn resolved(&self) -> Option<&ResolvedTy<'buf>>;

    fn is_resolved(&self) -> bool;
}

impl<'buf> private::Sealed for Option<Ty<'buf>> {}

impl<'buf> TyExt<'buf> for Option<Ty<'buf>> {
    fn raw(&self) -> Option<Option<&UnresolvedTy<'buf>>> {
        match self {
            None => Some(None),
            Some(Ty::Unresolved(ty)) => Some(Some(ty)),
            Some(Ty::Resolved(_)) => None,
        }
    }

    fn resolved(&self) -> Option<&ResolvedTy<'buf>> {
        self.as_ref().and_then(Ty::resolved)
    }

    fn is_resolved(&self) -> bool {
        self.as_ref().is_resolved()
    }
}

impl<'buf> private::Sealed for Option<&Ty<'buf>> {}

impl<'buf> TyExt<'buf> for Option<&Ty<'buf>> {
    fn raw(&self) -> Option<Option<&UnresolvedTy<'buf>>> {
        match self {
            None => Some(None),
            Some(Ty::Unresolved(ty)) => Some(Some(ty)),
            Some(Ty::Resolved(_)) => None,
        }
    }

    fn resolved(&self) -> Option<&ResolvedTy<'buf>> {
        self.and_then(Ty::resolved)
    }

    fn is_resolved(&self) -> bool {
        self.map(Ty::is_resolved).unwrap_or(false)
    }
}

impl<'buf> private::Sealed for Ty<'buf> {}

impl<'buf> TyExt<'buf> for Ty<'buf> {
    fn raw(&self) -> Option<Option<&UnresolvedTy<'buf>>> {
        match self {
            Ty::Unresolved(ty) => Some(Some(ty)),
            Ty::Resolved(_) => None,
        }
    }

    fn resolved(&self) -> Option<&ResolvedTy<'buf>> {
        try_match!(self, Self::Resolved(ty) => ty)
    }

    fn is_resolved(&self) -> bool {
        matches!(self, Self::Resolved(_))
    }
}

#[derive(Serialize, Debug, Clone, PartialEq, Eq, Hash)]
pub enum ResolvedTy<'buf> {
    Builtin(BuiltinClass),

    /// A `SELF_TYPE` occuring in the body of `enclosed`.
    SelfType {
        enclosed: Box<ResolvedTy<'buf>>,
    },

    /// A user-defined class.
    Class(
        /// The name of the class.
        #[serde(serialize_with = "serialize_bytes_as_string")]
        Cow<'buf, [u8]>,
    ),

    Function(FunctionTy<'buf>),

    /// The bottom type, a subclass of every type.
    ///
    /// A valid Cool program has no expression of this type;
    /// it only occurs as a result of a typing error.
    Bottom,

    /// An expression that cannot have a valid type in Col.
    ///
    /// How this is different from...
    /// - ...the top type: Untyped allows use in any covariant context
    /// - ...the bottom type: Untyped alllows use in any contravariant context
    ///
    /// An untyped expression can also be used in an invariant position.
    ///
    /// A valid Cool program has not expression of this type;
    /// it only occurs as a result of a typing error.
    Untyped,
}

impl CloneStatic<ResolvedTy<'static>> for ResolvedTy<'_> {
    fn clone_static(&self) -> ResolvedTy<'static> {
        match self {
            Self::Builtin(builtin) => ResolvedTy::Builtin(*builtin),
            Self::SelfType { enclosed } => ResolvedTy::SelfType {
                enclosed: Box::new(enclosed.clone_static()),
            },
            Self::Class(name) => ResolvedTy::Class(Cow::Owned(name.clone().into_owned())),
            Self::Function(ty) => ty.clone_static().into(),
            Self::Bottom => ResolvedTy::Bottom,
            Self::Untyped => ResolvedTy::Untyped,
        }
    }
}

impl Display for ResolvedTy<'_> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::Builtin(builtin) => builtin.fmt(f),
            Self::SelfType { enclosed } => write!(f, "SELF_TYPE[{}]", enclosed),
            Self::Class(name) => write!(f, "{}", slice_formatter(name)),
            Self::Function(ty) => ty.fmt(f),
            Self::Bottom => write!(f, "⊥"),
            Self::Untyped => write!(f, "{{invalid}}"),
        }
    }
}

/// A built-in class type.
#[derive(Serialize, Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum BuiltinClass {
    Int,
    String,
    Bool,

    /// The top type, a superclass of every type.
    Object,

    IO,
}

impl Display for BuiltinClass {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(
            f,
            "{}",
            match self {
                Self::Int => "Int",
                Self::String => "String",
                Self::Bool => "Bool",
                Self::Object => "Object",
                Self::IO => "IO",
            }
        )
    }
}

impl<'buf> From<BuiltinClass> for ResolvedTy<'buf> {
    fn from(builtin: BuiltinClass) -> ResolvedTy<'buf> {
        ResolvedTy::Builtin(builtin)
    }
}

/// A function type, used to represent method types.
#[derive(Serialize, Debug, Clone, PartialEq, Eq, Hash)]
pub struct FunctionTy<'buf> {
    pub params: Vec<ResolvedTy<'buf>>,
    pub ret: Box<ResolvedTy<'buf>>,
}

impl CloneStatic<FunctionTy<'static>> for FunctionTy<'_> {
    fn clone_static(&self) -> FunctionTy<'static> {
        FunctionTy {
            params: self.params.iter().map(ResolvedTy::clone_static).collect(),
            ret: Box::new(self.ret.clone_static()),
        }
    }
}

impl Display for FunctionTy<'_> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        for (i, param_ty) in self.params.iter().enumerate() {
            if i > 0 {
                write!(f, ", ")?;
            }

            write!(f, "{}", param_ty)?;
        }

        write!(f, " → {}", &self.ret)
    }
}

impl<'buf> From<FunctionTy<'buf>> for ResolvedTy<'buf> {
    fn from(ty: FunctionTy<'buf>) -> ResolvedTy<'buf> {
        ResolvedTy::Function(ty)
    }
}

impl<'buf> From<FunctionTy<'buf>> for Ty<'buf> {
    fn from(ty: FunctionTy<'buf>) -> Ty<'buf> {
        ResolvedTy::from(ty).into()
    }
}

pub trait HasTy<'buf> {
    fn ty(&self) -> Option<Cow<'_, Ty<'buf>>>;
}

pub trait HasExplicitTy<'buf> {
    fn explicit_ty(&self) -> Cow<'_, Ty<'buf>>;
}

impl<'buf, T: HasExplicitTy<'buf>> HasTy<'buf> for T {
    fn ty(&self) -> Option<Cow<'_, Ty<'buf>>> {
        Some(self.explicit_ty())
    }
}

impl<'buf> HasExplicitTy<'buf> for Ty<'buf> {
    fn explicit_ty(&self) -> Cow<'_, Ty<'buf>> {
        Cow::Borrowed(self)
    }
}

pub trait UnwrapResolvedTy<'buf>: HasTy<'buf> {
    fn unwrap_res_ty(&self) -> Cow<'_, ResolvedTy<'buf>>;
}

impl<'buf, T: HasTy<'buf>> UnwrapResolvedTy<'buf> for T {
    fn unwrap_res_ty(&self) -> Cow<'_, ResolvedTy<'buf>> {
        let resolved = match self
            .ty()
            .expect("unwrap_res_ty called on an uninferred type")
        {
            Cow::Borrowed(borrowed) => borrowed.resolved().map(Cow::Borrowed),
            Cow::Owned(owned) => match owned {
                Ty::Resolved(resolved) => Some(Cow::Owned(resolved)),
                Ty::Unresolved(_) => None,
            },
        };

        resolved.expect("unwrap_res_ty called on an unresolved type")
    }
}

mod private {
    pub trait Sealed {}
}
