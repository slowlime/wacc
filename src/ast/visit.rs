use super::*;

pub trait AstRecurse<'buf> {
    fn recurse<V: Visitor<'buf, Output = ()>>(&self, visitor: &mut V);
    fn recurse_mut<V: VisitorMut<'buf, Output = ()>>(&mut self, visitor: &mut V);
}

macro_rules! define_visitor {
    ($( $type:ident { $( $name:ident ( $arg:ident : $ty:ty ) );+ $(;)? } )+) => {
        pub trait Visitor<'buf>
        where
            Self: Sized,
        {
            type Output;

            $(
                $(
                    fn $name(&mut self, $arg: &$ty) -> Self::Output;
                )+
            )+
        }

        pub trait VisitorMut<'buf>
        where
            Self: Sized,
        {
            type Output;

            $(
                $(
                    fn $name(&mut self, $arg: &mut $ty) -> Self::Output;
                )+
            )+
        }

        pub trait DefaultVisitor<'buf>
        where
            Self: Sized,
        {
            $( define_visitor!(@ $type { $( $name ( $arg : &$ty ) => recurse; )+ } ); )+
        }

        impl<'buf, T> Visitor<'buf> for T
        where
            T: DefaultVisitor<'buf>,
        {
            type Output = ();

            $(
                $(
                    fn $name(&mut self, $arg: &$ty) {
                        <Self as DefaultVisitor<'buf>>::$name(self, $arg);
                    }
                )+
            )+
        }

        pub trait DefaultVisitorMut<'buf>
        where
            Self: Sized,
        {
            $( define_visitor!(@ $type { $( $name ( $arg : &mut $ty ) => recurse_mut; )+ } ); )+
        }

        impl<'buf, T> VisitorMut<'buf> for T
        where
            T: DefaultVisitorMut<'buf>,
        {
            type Output = ();

            $(
                $(
                    fn $name(&mut self, $arg: &mut $ty) {
                        <Self as DefaultVisitorMut<'buf>>::$name(self, $arg);
                    }
                )+
            )+
        }
    };

    (@ NonTerminal { $( $name:ident ( $arg:ident : $ty:ty ) => $recurse:ident; )+ }) => {
        $(
            fn $name(&mut self, $arg: $ty) {
                $arg.$recurse(self);
            }
        )+
    };

    (@ Terminal { $( $name:ident ( $arg:ident : $ty:ty ) => $recurse:ident; )+ }) => {
        $(
            #[allow(unused_variables)]
            fn $name(&mut self, $arg: $ty) {}
        )+
    };
}

define_visitor! {
    NonTerminal {
        // statements
        visit_program(program: Program<'buf>);
        visit_class(class: Class<'buf>);
        visit_feature(feature: Feature<'buf>);
        visit_method(method: Method<'buf>);
        visit_field(field: Field<'buf>);

        // expr
        visit_expr(expr: Expr<'buf>);
        visit_assignment(expr: Assignment<'buf>);
        visit_call(expr: Call<'buf>);
        visit_if(expr: If<'buf>);
        visit_while(expr: While<'buf>);
        visit_block(expr: Block<'buf>);
        visit_let(expr: Let<'buf>);
        visit_case(expr: Case<'buf>);
        visit_new(expr: New<'buf>);
        visit_bin_op(expr: BinOpExpr<'buf>);
        visit_un_op(expr: UnOpExpr<'buf>);
        visit_name_expr(expr: NameExpr<'buf>);

        // neither stmt nor expr
        visit_formal(formal: Formal<'buf>);
        visit_receiver(recv: Receiver<'buf>);
        visit_case_arm(arm: CaseArm<'buf>);
        visit_ty_name(ty_name: TyName<'buf>);
        visit_binding(binding: Binding<'buf>);
    }

    Terminal {
        visit_name(name: Name<'buf>);
        visit_int_lit(expr: IntLit);
        visit_string_lit(expr: StringLit<'buf>);
        visit_bool_lit(expr: BoolLit);
    }
}
