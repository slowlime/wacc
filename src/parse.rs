mod cursor;
mod lexer;
mod parser;
pub mod token;

pub use cursor::Cursor;
pub use lexer::{Lexer, LexerError};
pub use parser::{Parser, ParserError};
