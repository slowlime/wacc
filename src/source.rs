use std::fs;
use std::io;
use std::num::NonZeroUsize;
use std::path::{Path, PathBuf};

use elsa::FrozenVec;
use serde::Serialize;

#[derive(Serialize, Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub struct SourceId(NonZeroUsize);

#[derive(Default)]
pub struct SourceBuffer(FrozenVec<Vec<u8>>);

impl SourceBuffer {
    pub fn new() -> Self {
        Self(FrozenVec::new())
    }
}

pub struct Source<'buf> {
    buf: &'buf SourceBuffer,
    files: Vec<SourceFile<'buf>>,
}

#[derive(Debug)]
pub struct SourceFile<'buf> {
    id: SourceId,
    path: PathBuf,
    buf: &'buf [u8],
}

impl<'buf> Source<'buf> {
    pub fn new(buf: &'buf mut SourceBuffer) -> Self {
        // take a &mut to ensure exclusivity of the borrow

        Self {
            buf: &*buf,
            files: Vec::new(),
        }
    }

    pub fn load_from_string(&mut self, path: PathBuf, buf: Vec<u8>) -> SourceId {
        let idx = self.files.len();
        assert_eq!(self.buf.0.len(), idx);
        let id = SourceId((idx + 1).try_into().unwrap());

        self.buf.0.push(buf);
        let buf: &'buf [u8] = &self.buf.0[idx];

        self.files.push(SourceFile { id, path, buf });

        id
    }

    pub fn load(&mut self, path: PathBuf) -> io::Result<SourceId> {
        let buf = fs::read(&path)?;

        Ok(self.load_from_string(path, buf))
    }

    pub fn get(&self, id: SourceId) -> Option<&SourceFile<'buf>> {
        self.files.get(usize::from(id.0) - 1)
    }

    pub fn iter(&self) -> impl Iterator<Item = &SourceFile<'buf>> {
        self.files.iter()
    }
}

impl<'buf> SourceFile<'buf> {
    pub fn id(&self) -> SourceId {
        self.id
    }

    pub fn path(&self) -> &Path {
        &self.path
    }

    pub fn buf(&self) -> &'buf [u8] {
        self.buf
    }
}
