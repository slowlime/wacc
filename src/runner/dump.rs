use std::io::{self, Write};

use wacc::parse::token::{write_escaped_string, Symbol, SymbolCategory, Token, TokenValue};
use wacc::parse::LexerError;
use wacc::position::HasSpan;
use wacc::source::Source;

use super::config::LexerOutputFormat;

pub use wacc::ast::dump::{AstDumpError, AstDumpFormat, dump_ast};

pub fn dump_tokens<'buf, I>(
    format: LexerOutputFormat,
    source: &Source<'buf>,
    tokens: I,
    mut out: impl Write,
) -> io::Result<()>
where
    I: Iterator<Item = Result<Token<'buf>, LexerError>>,
{
    match format {
        LexerOutputFormat::Coolc => dump_tokens_coolc(source, tokens, out)?,
        LexerOutputFormat::Debug => writeln!(out, "{:#?}", tokens.collect::<Vec<_>>())?,
    }

    Ok(())
}

fn dump_tokens_coolc<'buf, I>(
    source: &Source<'buf>,
    tokens: I,
    mut out: impl Write,
) -> io::Result<()>
where
    I: Iterator<Item = Result<Token<'buf>, LexerError>>,
{
    let mut peekable = tokens.peekable();
    let src_id = peekable
        .peek()
        .expect("the lexer always returns at least one token")
        .as_ref()
        .map(|token| token.span.start.src)
        .unwrap_or_else(|err| err.span().start.src);
    let name = match src_id {
        Some(src_id) => format!(
            "{}",
            source
                .get(src_id)
                .expect("the span has a valid source id")
                .path()
                .display()
        ),
        None => "<unknown>".to_owned(),
    };
    write!(out, "#name ")?;
    write_escaped_string(name.as_bytes(), &mut out)?;
    writeln!(out)?;

    for token in peekable {
        let token = match token {
            Ok(Token {
                value: TokenValue::Eof,
                ..
            }) => continue,

            Ok(token) => token,

            Err(e) => {
                let span = e.span();
                write!(out, "#{} ERROR ", span.start.line)?;
                let msg = format!("{}", e);
                write_escaped_string(msg.as_bytes(), &mut out)?;
                writeln!(out)?;

                continue;
            }
        };

        write!(out, "#{} ", token.span.start.line)?;

        match token.value {
            TokenValue::Eof => unreachable!(),

            TokenValue::Int(i) => write!(out, "INT_CONST {}", i)?,

            TokenValue::Ident(id) => {
                if id.first().map(u8::is_ascii_uppercase).unwrap_or(false) {
                    out.write_all(b"TYPEID ")?;
                } else {
                    out.write_all(b"OBJECTID ")?;
                }

                out.write_all(&id)?;
            }

            TokenValue::Symbol(sym) => match sym.category() {
                // booleans enjoy special treatment by the reference lexer
                _ if [Symbol::True, Symbol::False].contains(&sym) => write!(
                    out,
                    "BOOL_CONST {}",
                    if sym == Symbol::True { "true" } else { "false" }
                )?,

                SymbolCategory::Keyword => write!(out, "{}", sym.as_str().to_ascii_uppercase())?,

                // weird edge-cases (since they're multibyte, I guess?)
                _ if sym == Symbol::ArrowLeft => write!(out, "ASSIGN")?,
                _ if sym == Symbol::Implies => write!(out, "DARROW")?,
                _ if sym == Symbol::LessEquals => write!(out, "LE")?,

                _ => write!(out, "'{}'", sym.as_str())?,
            },

            TokenValue::String(s) => {
                write!(out, "STR_CONST ")?;
                write_escaped_string(&s, &mut out)?;
            }
        }

        writeln!(out)?;
    }

    Ok(())
}
