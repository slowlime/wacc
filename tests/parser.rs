use std::borrow::Cow;
use std::path::Path;

use insta::assert_ron_snapshot;
use once_cell::unsync::OnceCell;
use paste::paste;
use pretty_assertions::assert_str_eq;

use wacc::ast::dump::{dump_ast, AstDumpFormat};
use wacc::parse::{Cursor, Lexer, Parser};
use wacc::source::{Source, SourceBuffer};

macro_rules! run_parser_test {
    ($filename:expr) => ({
        let mut source_buf = SourceBuffer::new();
        let mut source = Source::new(&mut source_buf);
        let code = include_bytes!(concat!("parser/", $filename, ".cl"));
        let source_id = source.load_from_string(
            Path::new("tests").join("parser").join($filename.to_string() + ".cl"),
            code.to_vec()
        );
        let source_file = source.get(source_id).unwrap();

        let cursor = Cursor::new(source_file);
        let lexer = Lexer::new(cursor);
        let parser = Parser::new(lexer);

        let parse_result = parser.parse();

        insta::with_settings!({
            description => String::from_utf8_lossy(code),
        }, {
            assert_ron_snapshot!(parse_result);
        });

        struct Dump<'a> {
            bytes: Cow<'a, [u8]>,
            string: OnceCell<String>,
        }

        impl PartialEq for Dump<'_> {
            fn eq(&self, other: &Dump) -> bool {
                self.bytes == other.bytes
            }
        }

        impl AsRef<str> for Dump<'_> {
            fn as_ref(&self) -> &str {
                self.string.get_or_init(|| {
                    String::from_utf8_lossy(&self.bytes).into_owned()
                })
            }
        }

        impl<'a> From<&'a [u8]> for Dump<'a> {
            fn from(bytes: &'a [u8]) -> Self {
                Self {
                    bytes: Cow::Borrowed(bytes),
                    string: OnceCell::new(),
                }
            }
        }

        impl From<Vec<u8>> for Dump<'_> {
            fn from(bytes: Vec<u8>) -> Self {
                Self {
                    bytes: Cow::Owned(bytes),
                    string: OnceCell::new(),
                }
            }
        }

        let coolc_actual = Dump::from(match parse_result {
            Ok(ast) => {
                let mut buf = vec![];
                dump_ast(
                    &source,
                    AstDumpFormat::Coolc {
                        ignore_types: true,
                    },
                    ast,
                    &mut buf
                )
                .unwrap();

                buf
            },

            Err(_) => b"ERR parsing failed\n".to_vec(),
        });

        let coolc_expected = Dump::from(
            include_bytes!(concat!("parser/", $filename, ".coolc-parse")).as_slice()
        );

        assert_str_eq!(coolc_actual, coolc_expected);
    });

    { $( $filename:ident ),+ $(,)? } => {
        $(
            paste! {
                #[test]
                fn [< test_ $filename >]() {
                    run_parser_test!(concat!(stringify!($filename)));
                }
            }
        )+
    }
}

run_parser_test! {
    fail_empty,
    pass_simple_class,
    pass_arith,
    pass_multi_class,
    pass_fields,
    pass_priority_assign,
    pass_priority_arith,
}
