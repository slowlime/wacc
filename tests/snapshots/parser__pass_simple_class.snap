---
source: tests/parser.rs
description: "class Main {\n    main(): Int {\n        0\n    };\n};\n"
expression: parse_result
---
Ok(Program(
  classes: [
    Class(
      name: TyName(Name(Spanned(
        value: "Main",
        span: Span(
          start: Position(
            src: Some(SourceId(1)),
            byte: 6,
            line: 1,
            col: 7,
          ),
          end: Position(
            src: Some(SourceId(1)),
            byte: 9,
            line: 1,
            col: 10,
          ),
        ),
      ))),
      inherits: None,
      features: [
        Method(Method(
          name: Name(Spanned(
            value: "main",
            span: Span(
              start: Position(
                src: Some(SourceId(1)),
                byte: 17,
                line: 2,
                col: 5,
              ),
              end: Position(
                src: Some(SourceId(1)),
                byte: 20,
                line: 2,
                col: 8,
              ),
            ),
          )),
          params: [],
          return_ty: TyName(Name(Spanned(
            value: "Int",
            span: Span(
              start: Position(
                src: Some(SourceId(1)),
                byte: 25,
                line: 2,
                col: 13,
              ),
              end: Position(
                src: Some(SourceId(1)),
                byte: 27,
                line: 2,
                col: 15,
              ),
            ),
          ))),
          body: Int(IntLit(Spanned(
            value: 0,
            span: Span(
              start: Position(
                src: Some(SourceId(1)),
                byte: 39,
                line: 3,
                col: 9,
              ),
              end: Position(
                src: Some(SourceId(1)),
                byte: 39,
                line: 3,
                col: 9,
              ),
            ),
          ))),
          span: Span(
            start: Position(
              src: Some(SourceId(1)),
              byte: 17,
              line: 2,
              col: 5,
            ),
            end: Position(
              src: Some(SourceId(1)),
              byte: 20,
              line: 2,
              col: 8,
            ),
          ),
          ty: Unresolved(Function(
            args: [],
            ret: Unresolved(Named(TyName(Name(Spanned(
              value: "Int",
              span: Span(
                start: Position(
                  src: Some(SourceId(1)),
                  byte: 25,
                  line: 2,
                  col: 13,
                ),
                end: Position(
                  src: Some(SourceId(1)),
                  byte: 27,
                  line: 2,
                  col: 15,
                ),
              ),
            ))))),
          )),
        )),
      ],
      span: Span(
        start: Position(
          src: Some(SourceId(1)),
          byte: 0,
          line: 1,
          col: 1,
        ),
        end: Position(
          src: Some(SourceId(1)),
          byte: 4,
          line: 1,
          col: 5,
        ),
      ),
      ty: Unresolved(Named(TyName(Name(Spanned(
        value: "Main",
        span: Span(
          start: Position(
            src: Some(SourceId(1)),
            byte: 6,
            line: 1,
            col: 7,
          ),
          end: Position(
            src: Some(SourceId(1)),
            byte: 9,
            line: 1,
            col: 10,
          ),
        ),
      ))))),
    ),
  ],
  span: Span(
    start: Position(
      src: Some(SourceId(1)),
      byte: 0,
      line: 1,
      col: 1,
    ),
    end: Position(
      src: Some(SourceId(1)),
      byte: 4,
      line: 1,
      col: 5,
    ),
  ),
))
